### ISSUE TITLE
Workflows fail when a default executor is not available, even if it has been overwritten in the workflow task

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#303]
I'm trying to run the cellpose_segmentation task on the FMI setup which has no gpu (and therefore, we don't have the gpu specified as a possible option of executors).

I used the new ability to edit the executor of a workflow task to cpu-low, which is available. But the workflow still fails early on with this error message:

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/middleware/cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/routing.py", line 68, in app
    await response(scope, receive, send)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/responses.py", line 165, in __call__
    await self.background()
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/background.py", line 43, in __call__
    await task()
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/starlette/background.py", line 26, in __call__
    await self.func(*self.args, **self.kwargs)
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/__init__.py", line 113, in submit_workflow
    output_dataset.meta = await process_workflow(
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/_parsl/__init__.py", line 211, in process_workflow
    final_task_pars_future = recursive_task_assembly(
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/_parsl/__init__.py", line 150, in recursive_task_assembly
    task_pars_depend_future = recursive_task_assembly(
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/_parsl/__init__.py", line 150, in recursive_task_assembly
    task_pars_depend_future = recursive_task_assembly(
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/_parsl/__init__.py", line 150, in recursive_task_assembly
    task_pars_depend_future = recursive_task_assembly(
  [Previous line repeated 1 more time]
  File "/opt/fmi/gliberal/fractal/miniconda3/envs/fractal-server-v1-a15/lib/python3.8/site-packages/fractal_server/app/runner/_parsl/__init__.py", line 141, in recursive_task_assembly
    raise ValueError(msg)
ValueError: this_task.executor='gpu' is not in data_flow_kernel.executors.keys()=dict_keys(['cpu-low', 'cpu-mid', 'cpu-high', '_parsl_internal'])
```

When checking the workflow, the relevant workflow task looks like this:

```
   {
      "workflow_id": 2,
      "task_id": 5,
      "order": 5,
      "meta": null,
      "args": {
        "args": {
          "labeling_channel": "A01_C01",
          "labeling_level": 2,
          "ROI_table_name": "well_ROI_table"
        },
        "meta": {
          "executor": "cpu-low"
        }
      },
      "id": 16,
      "task": {
        "name": "Cellpose Segmentation",
        "command": "/opt/fmi/gliberal/fractal/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.4.6/venv/bin/python /opt/fmi/gliberal/fractal/fractal-demos/examples-v1/server/FRAC
TAL_ROOT/.fractal/fractal-tasks-core0.4.6/venv/lib/python3.8/site-packages/fractal_tasks_core/cellpose_segmentation.py",
        "source": "pip:fractal-tasks-core==0.4.6",
        "input_type": "zarr",
        "output_type": "zarr",
        "default_args": {
          "labeling_channel": "A01_C01"
        },
        "meta": {
          "executor": "gpu",
          "parallelization_level": "image"
        },
        "id": 5
      }
    },

```

It appears the `recursive_task_assembly` of parsl uses the default executor, not the one provided to the workflow task. If this is just a parsl issue, then the new slurm executors should solve it. But would be good to have a test in place to check that a task that defaults to have a gpu executor can be used in a setup without gpu executors when the executor is overwritten in the workflow task to something that is available.

**Comment by tcompa** [comment]
Yes, this is a bug. In principle we should be loading the `WorkflowTask.executor` property (see https://github.com/fractal-analytics-platform/fractal-server/blob/main/fractal_server/app/runner/_parsl/__init__.py#L133), which is correctly set to use `"executor": "cpu-low"` in your case.

I cannot say whether the new backend fixes it, but for sure a test is due.

**Comment by tcompa** [comment]
When we define a project&workflow (for server `1.0.0b1` and slurm backend) as in
```bash
LABEL="2"

###############################################################################
# THINGS TO BE CHANGED BY THE USER
# Adapt to settings to where you run the example
BASE_FOLDER_EXAMPLE=`pwd`/..
###############################################################################

# Set useful variables
PRJ_NAME="proj-$LABEL"
DS_IN_NAME="input-ds-$LABEL"
DS_OUT_NAME="output-ds-$LABEL"
WF_NAME="Workflow $LABEL"

# Set cache path and remove any previous file from there
export FRACTAL_CACHE_PATH=`pwd`/".cache"
rm -rv ${FRACTAL_CACHE_PATH}

# Define/initialize empty project folder and temporary file
PROJ_DIR=`pwd`/tmp_${LABEL}
rm -r $PROJ_DIR
mkdir $PROJ_DIR

###############################################################################
# IMPORTANT: modify the following lines so that they point to absolute paths
INPUT_PATH=$BASE_FOLDER_EXAMPLE/images/10.5281_zenodo.7059515
OUTPUT_PATH=${PROJ_DIR}/output
###############################################################################

# Create project
OUTPUT=`fractal --batch project new $PRJ_NAME $PROJ_DIR`
PRJ_ID=`echo $OUTPUT | cut -d ' ' -f1`
DS_IN_ID=`echo $OUTPUT | cut -d ' ' -f2`
echo "PRJ_ID: $PRJ_ID"
echo "DS_IN_ID: $DS_IN_ID"

# Update dataset name/type, and add a resource
fractal dataset edit --name "$DS_IN_NAME" -t image --read-only $PRJ_ID $DS_IN_ID
fractal dataset add-resource -g "*.png" $PRJ_ID $DS_IN_ID $INPUT_PATH

# Add output dataset, and add a resource to it
DS_OUT_ID=`fractal --batch project add-dataset $PRJ_ID "$DS_OUT_NAME"`
echo "DS_OUT_ID: $DS_IN_ID"

fractal dataset edit -t zarr --read-write $PRJ_ID $DS_OUT_ID
fractal dataset add-resource -g "*.zarr" $PRJ_ID $DS_OUT_ID $OUTPUT_PATH

# Create workflow
WF_ID=`fractal --batch workflow new "$WF_NAME" $PRJ_ID`
echo "WF_ID: $WF_ID"

# Mapping of tasks to their IDs (until we add the by-name addressing)
#    "Create OME-ZARR structure": 1,
#    "Yokogawa to Zarr": 2,
#    "Replicate Zarr structure": 3,
#    "Maximum Intensity Projection": 4,
#    "Cellpose Segmentation": 5,
#    "Measurement": 6,
#    "Illumination correction": 7,
#    "Napari workflows wrapper": 8,
#    "Create OME-ZARR structure (multiplexing)": 9

# Add tasks to workflow
fractal workflow add-task $WF_ID 1 --args-file Parameters/args_create_ome_zarr.json
fractal workflow add-task $WF_ID 2
fractal workflow add-task $WF_ID 3
fractal workflow add-task $WF_ID 4
fractal workflow add-task $WF_ID 5 --args-file Parameters/args_cellpose_segmentation.json

# Look at the current workflows
fractal workflow show $WF_ID
echo
```

and then modify a workflow task as
```
$ fractal workflow edit-task 1 5  --json-file Parameters/cellpose_segmentation.json
```
where
```
$ cat Parameters/cellpose_segmentation.json 
{
   "args": {
        "labeling_level": 2
    },
    "meta": {
        "executor": "cpu-low"
    }
}
```
and if we then apply the workflow with `#fractal workflow apply -o $DS_OUT_ID -p $PRJ_ID $WF_ID $DS_IN_ID
`, then the task is submitted to the correct executor. This we can confirm by looking at the `partition` field in the submission script
```
#!/bin/sh
#SBATCH --output=/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/fractal-demos/examples-v1/server/artifacts/workflow_000001_job_000001/slurmpy.stdout.%j.log
#SBATCH --partition=main
#SBATCH --mem=1g
#SBATCH --nodes=1
#SBATCH --job-name Cellpose_Segmentation
export CFUT_DIR=/net/nfs4/pelkmanslab-fileserver-common/data/homes/fractal/fractal-demos/examples-v1/server/artifacts/workflow_000001_job_000001
srun /data/homes/fractal/.conda/envs/fractal-server-v1/bin/python -m cfut.remote FMwWTeSpyEyLiXQBH7ZxIG2xf8EfmcSx
```

This suggests that the current version does not have this bug any more.

**Comment by jluethi** [comment]
Should it make a difference whether we have the executor in the parameter of the task when it's added or when it's modified? Why did you add it with the parameters and then modify it with the same parameters above?

**Comment by jluethi** [comment]
I then manually did the `fractal workflow apply -o 2 -p 1 1 1 --worker-init "$WORKER_INIT"` and now my workflow task looks like this:

```
{
  "workflow_id": 1,
  "task_id": 5,
  "order": 4,
  "meta": {
    "executor": "cpu-low"
  },
  "args": {
    "args": {
      "labeling_channel": "A01_C01",
      "labeling_level": 3,
      "ROI_table_name": "well_ROI_table",
      "label_name": "nuclei"
    },
    "meta": {
      "executor": "cpu-low"
    },
    "labeling_channel": "A01_C01",
    "labeling_level": 3,
    "ROI_table_name": "well_ROI_table",
    "label_name": "nuclei"
  },
  "id": 5,
  "task": {
    "name": "Cellpose Segmentation",
    "command": "/opt/fmi/gliberal/fractal/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.5.1/venv/bin/python /opt/fmi/gliberal/fractal/fractal-demos/examples-v1/server/FRACTAL_ROOT/.fractal/fractal-tasks-core0.5.1/venv/lib/python3.8/site-packages/fractal_tasks_core/cellpose_segmentation.py",
    "source": "pip:fractal-tasks-core==0.5.1",
    "input_type": "zarr",
    "output_type": "zarr",
    "default_args": {
      "labeling_channel": "A01_C01"
    },
    "meta": {
      "executor": "gpu",
      "parallelization_level": "image"
    },
    "id": 5
  }
}
```

Now it works. And now I understand what broke before: It put the args & meta into an "args" keyword when I used it in add-task.

So, the old way of adding both args & meta as part of adding a task doesn't work anymore then, correct? Is there a way to still do it in one as part of fractal workflow add-task?

**Comment by tcompa** [comment]
Yes, there is a difference (maybe counterintuitive one, but there is one). Let's look at the help messages:

```
$ fractal workflow add-task -h
usage: fractal workflow add-task [-h] [--order ORDER] [--args-file ARGS_FILE] id task_id
positional arguments:
  id                    Workflow id
  task_id               Task id
optional arguments:
  -h, --help            show this help message and exit
  --order ORDER         Order of this task within the workflow's task list
  --args-file ARGS_FILE
                        Path to a json serialised file containing the arguments ovverrides of the task

$ fractal workflow edit-task -h
usage: fractal workflow edit-task [-h] [--json-file JSON_FILE] id workflow_task_id
positional arguments:
  id                    Workflow id
  workflow_task_id      Workflow task Id, the Id of a task inside the list of tasks
optional arguments:
  -h, --help            show this help message and exit
  --json-file JSON_FILE
                        Path to a json serialised file containing the attribute ovverrides of the task
```

`ARGS_FILE` and `JSON_FILE` have different meanings. The former is only for `args`, while the latter is both for `args` and `meta`. The first one should have no special key, while the second should have `args` and/or `meta` keys.

In fact in your logs you see:
```
"args": {
        "args": {
          "labeling_channel": "A01_C01",
          "labeling_level": 3,
          "ROI_table_name": "well_ROI_table",
          "label_name": "nuclei"
        },
```
which is incorrect. This would mean that the task has an argument called `args` which takes value `{"labeling_channel": "A01_C01", ...}`.

---

TL;DR
In the current version, the `add-task` can only modify the `args`, and not `meta`. To get it work with the current version, you should only write the args in `ARGS_FILE`, namely:
```JSON
{
          "labeling_channel": "A01_C01",
          "labeling_level": 3,
          "ROI_table_name": "well_ROI_table",
          "label_name": "nuclei"
        }
```

We  can then make things more uniform

**Comment by jluethi** [comment]
Ok, thanks!
